rfc/plugins.md

```diff
@@ -0,0 +1,343 @@
+For example, a remote plugin that exposes its integration through a gRPC server would also need a gRPC transport plugin on the service so that communication between the two can be configured correctly. 
+Other transport plugins, such as Docker co-process, Functions, HTTP, and so-on could also be developed.
+
+An extension point looks like the following:
+
+```kotlin
+// Definition of the PluginStage extension point.
+// The annotation defined here also informs the Spinnaker
+// plugin framwework which runtimes are acceptable. This
+// annotation would additionally be used for other metadata
+// Spinnaker needs to know atop whatever PF4J offers out of
+// the box.
+@SpinnakerExtension(allowRemote = true)
+class PipelineStageExtension : ExtensionPoint {
+  // ...
+}
+```
+
+It is highly recommended to read the documentation of PF4J to fully understand the scope of this proposal: This proposal will not go in depth on PF4J internals or taxonomy.
+
+### Front50 as the Source of Truth
+Operators will need the ability to understand, at a glance, what plugins are installed and enabled or disabled, their versions, as well as any other metadata about a plugin. 
+Whether plugins are implemented as in-process or not, Front50 should be used to track and expose a single source of truth of what plugins are installed in any particular service, as well as which ones are enabled or disabled.
+
+The Front50 plugin registry will also be exposed through Gate, and can be consumed by Deck to load the correct plugins on application startup.
+
+Plugins can also be enabled or disabled at runtime without redeploying a service. 
+This flag is sourced from Front50, which all services will be listening for change events from. Whether this change event is published or polled for initially is unknown.
+
+Halyard configs can be used to inform Spinnaker of desired state, but whatever state is defined in Front50 will be used as the canonical source for plugins. 
+When a service starts up, it will ask Front50 what plugins need to be installed and which of those needs to enabled or disabled. Halyard’s role will only go as far as updating Front50’s repository of plugin metadata.
+
+Using Front50 as a source of truth will require extension of PF4J.
+
+```proto
+syntax = "proto3";
+
+// PluginManifest is used during registration of a plugin
+message PluginManifest {
+	// The name of the plugin, which is used as a fully 
+	// qualified identifier, including a namespace and 
+	// plugin name, ex: "netflix/fast-properties"
+	string name = 1;
+
+	// A short description of the plugin
+	// ex: "Provides rollout safety of runtime application 
+	// config via pipelines"
+	string description = 2;
+
+	// The name of the plugin author, ex: "Netflix"
+	string provider_name = 3;
+
+	// The semver plugin version, ex: "1.0.0"
+	string version = 4;
+
+	// A list of plugin dependencies. Initially this will be
+	// informational only. ex: "netflix/somedep>=1.x"
+	repeated string dependencies = 5;
```

Comment:
going off the example, I'm assuming that would be referencing a registry of some sort? If not, how are plugin resources referenced (eg the location of the jars in an artifact store)? 

Also, do we want to make any distinction between dependencies for different services (eg Orca needs x.jar, while Deck needs y.js, z.png)? or is each service responsible for downloading all of the plugin resources? I think we could also register a plugin multiple times if it spans multiple services, but that leads to complexity

Reply:
riffing off this some more, we may want to add a type as well. This is more related to Deck, but I can see Deck needing to load js and css files in different ways, but that hasn't really been though out too much. 

I'm open to discussing this in another design doc, I'm not sure how much we want to grow the scope of this design, but a plugin frontend component is pretty important piece.

Reply:
Taxonomy thing: I've been trying to say that the plugin registry is Front50, in that plugins are registered into it. Plugin repositories would be where plugins are stored for download. Does this seem a reasonable distinction?

So, I think the `dependencies` property doesn't need to specify the repository that a plugin is found in. Kork should be capable of configuring a list of 1+n repositories that it will look in (ordered) to find a plugin (in this example, `netflix/somedep>=1.x`). Check out [pf4j-update](https://github.com/pf4j/pf4j-update) for inspiration here. Similar to the `pf4j-spring` integration, I don't see this `pf4j-update` library totally hitting the right mark for us, but you get the idea.

As for distinction of dependencies: Hard no from me, dude. 'No' means it'll be harder for us to develop, but I believe allowing plugin developers and operators alike to reason about their plugin as a single thing - even if it's packaged and distributed as many artifacts - is going to be important. Example:

`netflix/somedep` has Clouddriver, Orca, and Deck components. If we were building this plugin with Gradle, I'd expect 3 artifacts: `netflix/somedep-clouddriver-1.0`, `netflix/somedep-orca-1.0` and `netflix/somedep-deck-1.0`. When a clouddriver is configured to use `netflix/somedep` the download resolver would know to look for a `netflix/somedep-clouddriver` artifact. This isn't too far off from how Spinnaker service modules are written and published today.

Be interested for @christopherthielen to chime in on how Deck would need multiple artifacts (as you point out). Thoughts?

Reply:
the registry/ repository distinction makes sense 👍 

Ah, I see, I was missing the part about Kork configuring repositories to look at, but that makes total sense and is a common pattern 👍 

I'm happy with that tradeoff around distinction of dependencies, one of the main points of this project is to lower the barrier of entry and requiring plugin creators to know which service needs which dependency is asking a lot. I was thinking about lean binaries, only loading what you need, but I'd prefer to optimize for plugin developer experience.

Reply:
One of the things @christopherthielen brought up in the UI design doc is SRI (https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity). It would be nice to have this included for Deck resources to validate what the user is getting is correct.